Adding more detail to 'Why is Materialize unresponsive or slow?' troubleshooting docs (#22743)
MaterializeInc/materialize: doc/user/content/manage/troubleshooting.md

From the results of this query we can see that most of the elapsed time of the dataflow is caused by the time it takes to maintain the updates in one of the arrangements of the dataflow of the materialized view.

## Why is Materialize unresponsive or slow?
Slow responses can be the result of a number of factors, which we can help you understand and address.

### Indexes

> If you haven’t already, consider creating an [index](https://materialize.com/docs/sql/create-index/) for your query. The [optimizations](https://materialize.com/docs/transform-data/optimization/) page in our docs has more advanced guidance on building an efficient index. Indexes maintain a query’s up-to-date results in memory within a [cluster](https://materialize.com/docs/get-started/key-concepts#clusters), which allows future queries to return quickly.

Reviewer: This is good, and I'd like to think about if MVs should be mentioned alongside indexes here too. This is especially true depending on the use patterns of the query and if other clusters would like to consume the query/index or not. The "When to Use Indexes and Materialized Views" blog post is great below. I think it's good as is, though, this is more of a musing.

If you’re planning to run a query regularly and expect it to be fast, an index is highly recommended.

Some tips for troubleshooting using an index:

- Start by querying the [EXPLAIN](https://materialize.com/docs/sql/explain/) plan. This will tell you the query plan that Materialize will execute for your query.
- Once you add an index, re-issue the explain plan to ensure the index was actually used.

> - If you have already created an index, and it’s not showing up in the explain plan, you may be running the query in a different cluster than where the index resides. An index caches the results in memory on the cluster on which you create it. You need to execute your query on that same cluster in order to utilize the index.

Reviewer: This could also be because the index columns don't exactly match a predicate. Many people will come in expecting prefix matches to work but they don't with us. Correct as is, but this could also be a common reason.

> If you do not want to use an index, consider the size of the underlying dataset on which you’re querying. Without an index, the query is computed on the fly, so larger data sizes translate to more on-the-fly computation, which can lead to longer latencies.

Reviewer: "If you do not want to use an index" why would this be? Presumably only because of cluster mem/cpu use. Ok, so: an index takes X memory always to maintain computation. An identical non-indexed query presumably also takes around X memory but then releases it all. But the maximum consumed memory should be roughly the same, and the cpu use will go way up for non-indexed queries. I'd like more justification for this paragraph. Can it be removed, or are there some additional words to add describing why folks might not want to add an index?

For more on when to use an index, check out our blog post [When to Use Indexes and Materialized Views](https://materialize.com/blog/views-indexes/).

### Isolation level

> Isolation level is a way to specify how you want your query to behave with respect to the other queries, source ingestion, and index & materialize view updates happening at the same time within your Materialize region. All isolation levels at Materialize guarantee you’re getting a consistent snapshot of the data - we provide the effect that all transactions are running in serial order and you will not see partially-committed transaction data.

Reviewer: Probably worth linking https://materialize.com/docs/get-started/isolation-level/#choosing-the-right-isolation-level somewhere here which has a lot of this text. Joe and I were discussing recently in https://github.com/MaterializeInc/materialize/pull/22693#issuecomment-1781788084 whether we should duplicate information or try to link to a canonical place. A troubleshooting guide might be a good place to duplicate so users don't have to click to another doc. But copying some of the wording from there or explicitly quote linking it might be a good practice.

Author: Yeah hmm I agree. After thinking on it, I might punt on adding this isolation level section to the troubleshooting docs for now and rethinking how we can edit that page first, then make this guide quote what we put there.

The existing isolation level docs are great, but I think we need to add even more high-level content there about the "so what" and why users might care about these tradeoffs, perhaps even with some real world business/product examples.

> The default isolation level is Strict Serializable, which ensures that for each query you execute, you will always see results that are at least as fresh as previous runs of the query. You get the most up to date correct result each time. The tradeoff with Strict Serializable is that you may need need to wait for data ingested and updates made at a previous time to propagate through materialized views and indexes before you can run your next query, resulting in your queries being occasionally blocked.

Reviewer: ```suggestion
The default isolation level is Strict Serializable, which ensures that for each query you execute, you will always see results that are at least as fresh as previous runs of the query. You get the most up-to-date, correct result each time. The tradeoff with Strict Serializable is that you may need to wait for data to be ingested and updates made at a previous time to propagate through materialized views and indexes. This results in your queries being occasionally blocked.
```

> at least as fresh as previous runs of the query

It's also at least as fresh as any other query. Is that worth saying?

> you may need to wait for data to be ingested

This is currently the case because of the 300ms upper propagation delay? Or are you referring to something else?

> An alternative isolation level is Serializable. With Serializable level, the results are still guaranteed to be correct, but may be from an earlier point in time. As a result, queries do not have to wait for recent updates and data ingestion to propagate, thus reducing the likelihood of your query being blocked. In Serializable, queries that involve a single object should never block.

Reviewer: ```suggestion
An alternative isolation level is Serializable. In Serializable, results are still guaranteed to be correct, but may be from an earlier point in time. As a result, queries do not have to wait for recent updates and data ingestion to propagate, thus reducing the likelihood of your query being blocked. In Serializable, auto-commit queries that involve a single object should never block.
```

Need to mention auto-commit here because otherwise we might bring in nearby objects.

What you can do: if you...

- care more about latency than freshness, and would prefer to receive fast, stale results \
AND
- are ok with queries "traveling back in time" with respect to other queries in order to serve those fast, stale results

then you can switch your isolation level to Serializable.

To change your session's isolation level to Serializable, run
```

> SET transaction_isolation TO 'SERIALIZABLE';

Author: Todo: add the below once default session vars per role is in GA

> To make this the value the default for your role, so that it applies every time you connect, run
> 
> `ALTER ROLE <your role> SET transaction_isolation = 'SERIALIZABLE';`
